modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/GridNearTransactionalCache.java

```diff
@@ -627,6 +652,166 @@ else if (!F.isEmpty(req.keys()))
+
+                                ctx.mvcc().addFuture(fut, fut.futureId());
+
+                                unlockFuts.put(req, fut);
+                            }
+                        }
+
+                        // Remove candidate from local node first.
+                        GridCacheMvccCandidate rmv = entry.removeLock();
+
+                        if (rmv != null) {
+                            if (!rmv.reentry()) {
+                                if (ver != null && !ver.equals(rmv.version()))
+                                    throw new IgniteCheckedException("Failed to unlock " +
+                                        "(if keys were locked separately, " +
+                                        "then they need to be unlocked separately): " + keys);
+
+                                if (!primary.isLocal()) {
+                                    req.addKey(entry.key(), ctx);
+
+                                    unlockFuts.get(req).addKey(key);
+                                }
+                                else
+                                    locKeys.add(cacheKey);
+
+                                if (log.isDebugEnabled())
+                                    log.debug("Removed lock (will distribute): " + rmv);
+                            }
+                            else if (log.isDebugEnabled())
+                                log.debug("Current thread still owns lock (or there are no other nodes)" +
+                                    " [lock=" + rmv + ", curThreadId=" + Thread.currentThread().getId() + ']');
+                        }
+                    }
+
+                    assert !topVer.equals(AffinityTopologyVersion.NONE) || cand == null;
+
+                    if (topVer.equals(AffinityTopologyVersion.NONE))
+                        topVer = ctx.affinity().affinityTopologyVersion();
+
+                    ctx.evicts().touch(entry, topVer);
+
+                    break;
+                }
+                catch (GridCacheEntryRemovedException ignore) {
+                    if (log.isDebugEnabled())
+                        log.debug("Attempted to unlock removed entry (will retry): " + entry);
+                }
+            }
+        }
+
+        if (ver == null)
+            return;
+
+        for (Map.Entry<ClusterNode, GridNearUnlockRequest> mapping : map.entrySet()) {
+            ClusterNode n = mapping.getKey();
+
+            GridDistributedUnlockRequest req = mapping.getValue();
+
+            if (n.isLocal())
+                dht.removeLocks(ctx.nodeId(), req.version(), locKeys, true);
+            else if (!F.isEmpty(req.keys())) {
+                ctx.io().send(n, req, ctx.ioPolicy());
+            }
+        }
+
+        ArrayList<IgniteTxKey> notUnlockedKeys = null;
+
+        for (SavepointUnlockFuture fut : unlockFuts.values()) {
+            try {
+                fut.get();
+            } catch (IgniteCheckedException e) {
```

Comment:
1) exception can be due to some other reasons(not only timeout), so u must throw checked exception.
2)after you got exception, you must break the cycle and stop other features.

Reply:
I gather info about all failed futures and throw one exception for them all together.

Reply:
yeah, and its incorrect. See org.apache.ignite.internal.processors.cache.distributed.near.GridNearOptimisticTxPrepareFuture.MiniFuture#onResult(org.apache.ignite.internal.processors.cache.distributed.near.GridNearTxPrepareResponse) for example